andes/routines/glodf.py

```diff
@@ -0,0 +1,244 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Mon Apr 24 19:59:15 2023
+
+@author: rvaug
+"""
+
+import andes
+import pandas as pd
+import numpy as np
+
+class GLODF:
+    """
+    Computes the Global Line Outage Distribution Factors (GLODF) for a power system.
+
+    Parameters
+    system : The power system model.
+    """
+    def __init__(self, system):
+        self.system = system
+        self.Bs = self.get_branch_susceptance()
+        self.Ar = self.get_reduced_incidence()
+        self.Bn = self.get_reduced_nodal_susceptance()
+
+    def get_branch_susceptance(self):
+        """
+        Returns the branch susceptance matrix.
+
+        Returns
+        -------
+        numpy.ndarray : Bs
+            The L x L diagonal matrix of branch susceptances.
+        """
+        return np.diag(1/self.system.Line.x.v)
+    
+    def get_reduced_incidence(self):
+        """
+        Returns the reduced incidence matrix.
+
+        Returns
+        -------
+        numpy.ndarray : Ar
+            The L x (N-s) reduced incidence matrix where 's' is the number of 
+            slack buses.
+        """
+        num_line = self.system.Line.n
+        num_bus = self.system.Bus.n
+        
+        A = np.zeros((num_line, num_bus))
+        
+        for l in range(num_line):
+            A[l,self.system.Line.bus1.v[l]-1] = -1
+            A[l,self.system.Line.bus2.v[l]-1] =  1
+        
+        # delete all slack rows as required
+        Ar = np.delete(A, np.asarray(self.system.Slack.bus.v)-1, axis=1)
```

Comment:
This line will fail if `self.system.Slack.bus.v` contains non-numerical values. `Slack.bus.v` contains references to bus `idx`es, which are not guaranteed numerals.

You can call `self.system.Bus.idx2uid(self.system.Slack.bus.v)` to retrieve 0-based index of the Slack bus id.